install-shoulda-and-factory-girl

From: Bryan Larsen <bryan@larsen.st>

# Unit Testing

They say that an application without tests is a broken application, so
lets add some unit tests.

When starting out with Rails testing, you are faced with a bewildering
array of choices.   It's outside the scope of this tutorial to guide
you through these choices, but I will give you one piece of advice:
any tests are better than no tests.   Do not get caught up in
"analysis paralysis" -- don't spend so much time trying to analyze
which choices are right for you that you don't do any testing at all.
If you change your mind later, it isn't that hard to translate tests
from one environment to another.

Also be aware that choices that are good for one area are not
necessarily good for another.  You might use a factory for unit
testing, but fixtures for integration testing, for example.

For this tutorial we're going to go with a more mainstream choice:
Test::Unit with Shoulda for the test DSL and FactoryGirl with Faker to
create test data.

I'm going to give a special shout out here to
[IRT](https://github.com/ddnexus/irt),  which is written by the same
man who ported Hobo to Rails3: Domizio Demichelis.  It offers a unique
paradigm that makes it easy to simultaneously develop code and tests.
Highly recommended.

We aren't going to use IRT in this tutorial, but there's nothing
stopping you from writing both Test::Unit and IRT tests in the same
application.

When we initially generated our application, we didn't customize our
test framework.   So let's revisit that decision.

    $ hobo generate test_framework shoulda --fixture_replacement=factory_girl
    $ bundle install
---
 Gemfile               |    2 ++
 Gemfile.lock          |   12 ++++++++++++
 config/application.rb |    7 +++++++
 3 files changed, 21 insertions(+)

diff --git a/Gemfile b/Gemfile
index 42df1de..42648b1 100644
--- a/Gemfile
+++ b/Gemfile
@@ -49,3 +49,5 @@ gem "jquery-ui-themes", "~> 0.0.4"
 
 gem "acts_as_list"
 gem "bluecloth"
+gem "shoulda", :group => :test
+gem "factory_girl_rails", :group => :test
diff --git a/Gemfile.lock b/Gemfile.lock
index 35fb9f9..60e085d 100644
--- a/Gemfile.lock
+++ b/Gemfile.lock
@@ -55,6 +55,11 @@ GEM
     erubis (2.7.0)
     execjs (1.4.0)
       multi_json (~> 1.0)
+    factory_girl (2.6.3)
+      activesupport (>= 2.3.9)
+    factory_girl_rails (1.7.0)
+      factory_girl (~> 2.6.0)
+      railties (>= 3.0.0)
     hike (1.2.1)
     hobo (2.0.0.pre8)
       dryml (= 2.0.0.pre8)
@@ -133,6 +138,11 @@ GEM
       railties (~> 3.2.0)
       sass (>= 3.1.10)
       tilt (~> 1.3)
+    shoulda (3.0.1)
+      shoulda-context (~> 1.0.0)
+      shoulda-matchers (~> 1.0.0)
+    shoulda-context (1.0.0)
+    shoulda-matchers (1.0.0)
     sprockets (2.2.2)
       hike (~> 1.2)
       multi_json (~> 1.0)
@@ -158,6 +168,7 @@ DEPENDENCIES
   acts_as_list
   bluecloth
   coffee-rails (~> 3.2.1)
+  factory_girl_rails
   hobo (= 2.0.0.pre8)
   hobo_bootstrap (= 2.0.0.pre8)
   hobo_bootstrap_ui (= 2.0.0.pre8)
@@ -167,6 +178,7 @@ DEPENDENCIES
   quiet_assets
   rails (= 3.2.11)
   sass-rails (~> 3.2.3)
+  shoulda
   sqlite3
   uglifier (>= 1.0.3)
   will_paginate!
diff --git a/config/application.rb b/config/application.rb
index 37588fb..c2516d1 100644
--- a/config/application.rb
+++ b/config/application.rb
@@ -17,6 +17,13 @@ module Agility
     config.hobo.dryml_only_templates = true
     # Hobo: the front subsite loads front.css & front.js
     config.assets.precompile += %w(front.css front.js)
+
+    config.generators do |g|
+      g.test_framework :shoulda, :fixtures => true
+      g.fallbacks[:shoulda] = :test_unit
+      g.fixture_replacement = :factory_girl_rails
+    end
+
     # Settings in config/environments/* take precedence over those specified here.
     # Application configuration should go into files in config/initializers
     # -- all .rb files in that directory are automatically loaded.
